cargo-sqlx/src/sqlite.rs

```diff
@@ -0,0 +1,167 @@
+// use sqlx::pool::PoolConnection;
+use sqlx::sqlite::SqliteRow;
+use sqlx::Connect;
+use sqlx::Executor;
+use sqlx::Row;
+use sqlx::SqliteConnection;
+// use sqlx::SqlitePool;
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+
+use crate::database_migrator::{DatabaseMigrator, MigrationTransaction};
+
+pub struct Sqlite {
+    db_url: String,
+    path: String,
+}
+
+impl Sqlite {
+    pub fn new(db_url: String) -> Self {
+        let path = crop_letters(&db_url, "sqlite://".len());
+        Sqlite {
+            db_url: db_url.clone(),
+            path: path.to_string(),
+        }
+    }
+}
+
+fn crop_letters(s: &str, pos: usize) -> &str {
+    match s.char_indices().skip(pos).next() {
+        Some((pos, _)) => &s[pos..],
+        None => "",
+    }
+}
+
+#[async_trait]
+impl DatabaseMigrator for Sqlite {
+    fn database_type(&self) -> String {
+        "Sqlite".to_string()
+    }
+
+    fn can_migrate_database(&self) -> bool {
+        true
+    }
+
+    fn can_create_database(&self) -> bool {
+        true
+    }
+
+    fn can_drop_database(&self) -> bool {
+        true
+    }
+
+    fn get_database_name(&self) -> Result<String> {
+        let split: Vec<&str> = self.db_url.rsplitn(2, '/').collect();
+
+        if split.len() != 2 {
+            return Err(anyhow!("Failed to find database name in connection string"));
+        }
+
+        let db_name = split[0];
+
+        Ok(db_name.to_string())
+    }
+
+    async fn check_if_database_exists(&self, _db_name: &str) -> Result<bool> {
+        use std::path::Path;
+        Ok(Path::new(&self.path).exists())
+    }
+
+    async fn create_database(&self, _db_name: &str) -> Result<()> {
+        use std::fs::OpenOptions;
+
+        println!("DB {}", self.path);
+
+        OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(&self.path)?;
+
+        Ok(())
+    }
+
+    async fn drop_database(&self, _db_name: &str) -> Result<()> {
+        std::fs::remove_file(&self.path)?;
```

Comment:
There is a C API for getting the file path from an existing database connection. We could expose that as a `#[doc(hidden)]` API on `SqliteConnection`.

Reply:
Will take a look. To be honest I am not quite happy with how I get the db name in postgres either, seems pretty fragile. Connection string can be pretty gnarly if you want to support all variants.